### PR TITLE
Addressed Comments from PR added save state

### PR DESCRIPTION
**Pull request by complextheory**

### CODE REVIEW

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/Block.kt [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-    var xOrO: String,
+@Parcelize
+data class Block(
+
+    var xOrO: String?,
     val id: Int,
     var isClicked: Boolean,
-    var playerClicked: Int
-)
+    var player: Int
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/Block.kt:16]: Changed variable name as suggested

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragment.kt [hunk 2]
```diff
@@ -1,32 +1,30 @@
...
+ * The Main Game Fragment That Handles UI Elements.
  */
 class GameFragment : Fragment() {
 
     private val viewModel: GameFragmentViewModel by viewModel()
 
     private lateinit var binding: FragmentGameBinding
 
-    private val blockItemClickListner = object: BlockClickListner {
+    private val blockItemClickListener = object: BlockClickListener {
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragment.kt:27]: Fixed Typo as suggested

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt [hunk 3]
```diff
@@ -1,53 +1,86 @@
...
-    private var blockList = mutableListOf<Block>()
-    private var player1BlockList = mutableListOf<Int>()
-    private var player2BlockList = mutableListOf<Int>()
-    private var drawGameBlockList = mutableListOf<Int>()
-    private var playerTurn = 1
-    var winner = 0
+class GameFragmentViewModel(private val savedState: SavedStateHandle) : ViewModel() {
+
+    var blockLiveData: MutableLiveData<List<Block>> =
+        savedState.getLiveData<List<Block>>(BLOCK_LIST_KEY).switchMap {
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:14]: Made Changes to Handle Config Change

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt [hunk 4]
```diff
@@ -68,44 +101,51 @@ class GameFragmentViewModel : ViewModel() {
...
+
+        return  (player1BlockList.contains(1) && player1BlockList.contains(2) && player1BlockList.contains(3)) ||
+                (player1BlockList.contains(4) && player1BlockList.contains(5) && player1BlockList.contains(6)) ||
+                (player1BlockList.contains(7) && player1BlockList.contains(8) && player1BlockList.contains(9)) ||
+                (player1BlockList.contains(1) && player1BlockList.contains(4) && player1BlockList.contains(7)) ||
+                (player1BlockList.contains(2) && player1BlockList.contains(5) && player1BlockList.contains(8)) ||
+                (player1BlockList.contains(3) && player1BlockList.contains(6) && player1BlockList.contains(9)) ||
+                (player1BlockList.contains(1) && player1BlockList.contains(5) && player1BlockList.contains(9)) ||
+                (player1BlockList.contains(3) && player1BlockList.contains(5) && player1BlockList.contains(7)) ||
+                (player2BlockList.contains(1) && player2BlockList.contains(2) && player2BlockList.contains(3)) ||
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:142]: Refactored Function as suggested

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt [hunk 5]
```diff
@@ -68,44 +101,51 @@ class GameFragmentViewModel : ViewModel() {
...
         playerTurn = 1
         counter = 1
+        updateSavedState()
         return blockList
     }
 
+    private fun updateSavedState() {
+        savedState[IS_GAME_IN_PROGRESS_KEY] = isGameInProgress
+        savedState[PLAYER_1_LIST_KEY] = player1BlockList
+        savedState[PLAYER_2_LIST_KEY] = player2BlockList
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:111]: Function to update saveState

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/TableAdapter.kt [hunk 6]
```diff
@@ -49,18 +46,19 @@ class TableAdapter(
...
+                binding.blockView.setBackgroundColor(Color.BLACK)
 
-                if (block.isClicked && block.playerClicked == 1){
-                    binding.blockView.setBackgroundColor(Color.BLUE)
-                }else if (block.isClicked && block.playerClicked == 2){
-                    binding.blockView.setBackgroundColor(Color.RED)
-                }else
-                    binding.blockView.setBackgroundColor(Color.BLACK)
+                if (block.isClicked){
+                    when (block.player) {
```
Comment (complextheory) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/TableAdapter.kt:55]: Cleaned up Binding as Suggested

File: TikTacToeGame/app/src/test/java/com/r4ziel/tiktactoegame/GameFragmentViewModelTest.kt [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+import org.junit.runners.JUnit4
+
+
+/**
+ * Created by Jarvis Charles on 5/28/23.
+ */
+
+@RunWith(JUnit4::class)
+class GameFragmentViewModelTest{
+
```
Comment (complextheory) [inline TikTacToeGame/app/src/test/java/com/r4ziel/tiktactoegame/GameFragmentViewModelTest.kt:20]: Added Unit Tests as Suggested

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/modules/ViewModelModule.kt [hunk 8]
```diff
@@ -8,5 +8,5 @@ import org.koin.dsl.module
  */
 
 val viewModelModule = module {
-    single { GameFragmentViewModel() }
+    single { GameFragmentViewModel(get()) }
```
Comment (droidclicketyclacker) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/modules/ViewModelModule.kt:11]: FYI: Koin has a new feature where you no longer have to add a get() for each constructor parameter. 

```suggestion
    singleOf(::GameFragmentViewModel)
```

File: TikTacToeGame/app/src/test/java/com/r4ziel/tiktactoegame/GameFragmentViewModelTest.kt [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        //Reset Values For Next Game
+        playerTurn = 1
+        counter = 1
+
+        return blocklist
+    }
+
+    @Test
+    fun gameFragmentViewModel_PlayerTurnIs1SavedStateIsSavedPlayerTurnIsUpdated() {
```
Comment (droidclicketyclacker) [inline TikTacToeGame/app/src/test/java/com/r4ziel/tiktactoegame/GameFragmentViewModelTest.kt]: Kotlin coding conventions indicate that we can use string names for tests to be more descriptive. Let's update all the test names according to Kotlin standards. See https://kotlinlang.org/docs/coding-conventions.html#property-names

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt [hunk 10]
```diff
@@ -68,44 +101,51 @@ class GameFragmentViewModel : ViewModel() {
...
-            (player2BlockList.contains(1) && player2BlockList.contains(4) && player2BlockList.contains(7)) ||
-            (player2BlockList.contains(2) && player2BlockList.contains(5) && player2BlockList.contains(8)) ||
-            (player2BlockList.contains(3) && player2BlockList.contains(6) && player2BlockList.contains(9)) ||
-            (player2BlockList.contains(1) && player2BlockList.contains(5) && player2BlockList.contains(9)) ||
-            (player2BlockList.contains(3) && player2BlockList.contains(5) && player2BlockList.contains(7))) {
-
-            winner = 2
-            return true
+        winner = if (player1BlockList.size == player2BlockList.size) {
+            2
```
Comment (droidclicketyclacker) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:130]: You can simplify this further:
```suggestion
        winner = 1
        if (player1BlockList.size == player2BlockList.size) winner = 2
```

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt [hunk 11]
```diff
@@ -1,53 +1,86 @@
...
     fun startGame() {
+        isGameInProgress = true
         blockLiveData.value = mutableListOf()
         isGameOverLiveData.value = false
         isDrawLiveData.value = false
         blockLiveData.postValue(generateBlocks())
+        updateSavedState()
     }
 
     fun updateTurn(block: Block) {
```
Comment (droidclicketyclacker) [inline TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:63]: Let's take a look at refactoring this method. There is some duplication between the player turns that can be simplified.

File: TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt (hunk 4 above)
Comment (droidclicketyclacker) [reply TikTacToeGame/app/src/main/java/com/r4ziel/tiktactoegame/GameFragmentViewModel.kt:142]: Nice formatting. Let's take a look at refactoring the logic. There's some duplication here between the player1 and player2 lists that can be simplified.

**Review by droidclicketyclacker — COMMENTED** [review]
Let's take a look at these few things in a follow up PR

**Pull request merged**